File: example-system/sim/sensor_node/filter.py
"""Moving-average filter (SN-REQ-003). Mirrors src/filter.c."""

from collections import deque


class MovingAverage:
    def __init__(self, window: int = 4) -> None:
        if window < 1:
            raise ValueError("window must be >= 1")
        self.window = window
        self._buf: deque[int] = deque(maxlen=window)

    def reset(self) -> None:
        self._buf.clear()

    def update(self, sample: int) -> int:
        """Push one sample and return the integer average of the samples seen so far."""
        self._buf.append(int(sample))
        total = sum(self._buf)
        n = len(self._buf)
        # C-style division: truncate toward zero, as src/filter.c does.
        q = abs(total) // n
        return q if total >= 0 else -q

    @property
    def count(self) -> int:
        return len(self._buf)

File: example-system/src/filter.c
#include "filter.h"

void filter_init(filter_t *f) { filter_reset(f); }

void filter_reset(filter_t *f) {
    for (uint8_t i = 0; i < FILTER_WINDOW; i++) f->buf[i] = 0;
    f->head = 0;
    f->count = 0;
}

int16_t filter_update(filter_t *f, int16_t sample) {
    f->buf[f->head] = sample;
    f->head = (uint8_t)((f->head + 1) % FILTER_WINDOW);
    if (f->count < FILTER_WINDOW) f->count++;
    int32_t sum = 0;
    for (uint8_t i = 0; i < f->count; i++) sum += f->buf[i];
    return (int16_t)(sum / f->count); /* C division truncates toward zero */
}
